Makefile

```diff
@@ -151,23 +168,36 @@ package-linux: GOOS:=linux
 package-linux: ARCHIVE_NAME=lifecycle-v$(LIFECYCLE_VERSION)+$(GOOS).x86-64
 package-linux:
-	@echo "> Writing descriptor file for $(GOOS)..."
-	mkdir -p $(GOOS_DIR)
-	echo "$${LIFECYCLE_DESCRIPTOR}" > $(GOOS_DIR)/lifecycle.toml
-
 	@echo "> Packaging lifecycle for $(GOOS)..."
-	tar czf $(BUILD_DIR)/$(ARCHIVE_NAME).tgz -C $(GOOS_DIR) lifecycle.toml lifecycle
+	$(GOCMD) run tools/packager/main.go -os $(GOOS) -launcherExePath $(GOOS_DIR)/lifecycle/launcher -lifecycleExePath $(GOOS_DIR)/lifecycle/lifecycle -lifecycleVersion $(LIFECYCLE_VERSION) -platformAPI $(PLATFORM_API) -buildpackAPI $(BUILDPACK_API) -outputPackagePath $(BUILD_DIR)/$(ARCHIVE_NAME).tgz
 
 package-windows: export LIFECYCLE_DESCRIPTOR:=$(LIFECYCLE_DESCRIPTOR)
 package-windows: GOOS:=windows
 package-windows: GOOS_DIR:=$(BUILD_DIR)/$(GOOS)
 package-windows: ARCHIVE_NAME=lifecycle-v$(LIFECYCLE_VERSION)+$(GOOS).x86-64
 package-windows:
 	@echo "> Writing descriptor file for $(GOOS)..."
-	mkdir -p $(GOOS_DIR)
-	echo "$${LIFECYCLE_DESCRIPTOR}" > $(GOOS_DIR)/lifecycle.toml
-
-	@echo "> Packaging lifecycle for $(GOOS)..."
-	tar czf $(BUILD_DIR)/$(ARCHIVE_NAME).tgz -C $(GOOS_DIR) lifecycle.toml lifecycle
+	$(GOCMD) run tools/packager/main.go -os $(GOOS) -launcherExePath $(GOOS_DIR)/lifecycle/launcher.exe -lifecycleExePath $(GOOS_DIR)/lifecycle/lifecycle.exe -lifecycleVersion $(LIFECYCLE_VERSION) -platformAPI $(PLATFORM_API) -buildpackAPI $(BUILDPACK_API) -outputPackagePath $(BUILD_DIR)/$(ARCHIVE_NAME).tgz
+
+docker-build-source-image-windows:
+	docker build -f tools/Dockerfile.windows --tag $(SOURCE_COMPILATION_IMAGE) --build-arg image_tag=$(WINDOWS_COMPILATION_IMAGE) --cache-from=$(SOURCE_COMPILATION_IMAGE) --isolation=process --quiet .
+
+docker-run-windows: docker-build-source-image-windows
+docker-run-windows:
+	@echo "> Running '$(DOCKER_CMD)' in docker windows..."
+	@docker volume rm -f lifecycle-out
+	docker run -v lifecycle-out:c:/lifecycle/out -e LIFECYCLE_VERSION -e PLATFORM_API -e BUILDPACK_API -v gopathcache:c:/gopath -v '\\.\pipe\docker_engine:\\.\pipe\docker_engine' --isolation=process --rm $(SOURCE_COMPILATION_IMAGE) $(DOCKER_CMD)
+	docker run -v lifecycle-out:c:/lifecycle/out --rm $(SOURCE_COMPILATION_IMAGE) tar -cf- out | tar -xf-
+	@docker volume rm -f lifecycle-out
+
+docker-build-source-image-linux:
+	docker build --tag $(SOURCE_COMPILATION_IMAGE) -f tools/Dockerfile.linux --build-arg image_tag=$(LINUX_COMPILATION_IMAGE) --cache-from=$(SOURCE_COMPILATION_IMAGE) --quiet .
+
+docker-run-linux: docker-build-source-image-linux
+docker-run-linux:
+	@echo "> Running '$(DOCKER_CMD)' in docker linux..."
+	@docker volume rm -f lifecycle-out
+	docker run -v lifecycle-out:/lifecycle -e LIFECYCLE_VERSION -v gopathcache:/go -v /var/run/docker.sock:/var/run/docker.sock --rm $(SOURCE_COMPILATION_IMAGE) $(DOCKER_CMD)
+	docker run -v lifecycle-out:/lifecycle --rm $(SOURCE_COMPILATION_IMAGE) tar -cf- out | tar -xf-
+	@docker volume rm -f lifecycle-out
```

Comment:
This is not currently used but is functionally equivalent to `docker-run-windows`. It could be incorporated into in `build-linux-lifecycle` to avoid the uncached `apk add` call and would reduce setup setups in the GHA  `build.yml`

Reply:
It seems like `docker-run-linux` has some issues. It can't run `make acceptance` b/c the image is missing the `docker` CLI.

Also, when I tried `DOCKER_CMD="make unit" make docker-run-linux` I got the following error
```
docker run -v lifecycle-out:/lifecycle --rm lifecycle-img tar -cf- out | tar -xf-
tar: out: No such file or directory
```
It seems like we only want to `docker run -v lifecycle-out:/lifecycle --rm $(SOURCE_COMPILATION_IMAGE) tar -cf- out | tar -xf-` after particular tasks?

Reply:
> It seems like `docker-run-linux` has some issues. It can't run `make acceptance` b/c the image is missing the `docker` CLI.

Right, I believe that's due to the `make docker-run-linux` (container)-> `make acceptance` -> `go test acceptance_test.go` -> `make build-linux` -> `docker run ... go build` chain. I'm interested in how this could best work:
1.) Changing `make build-linux` to be more like - though not quite equal to - `make build-windows` (which now switches based on OS to either `build-windows-on-windows` or `build-windows-on-posix`). Maybe the targets for `build-linux` would be something like `build-linux-on-linux`/`build-linux-on-docker` depending on OS. The target `build-linux-on-docker` would essentially just do `docker-run-linux DOCKER_CMD="make build-linux"` I think builds would then be faster since we could remove the `apk add build-base` for every call and GOPATH could be cached.
2.) or, adding `docker` to the `tools/Dockerfile.linux` image to support DIND, though we'd miss out on speedup from of removing `apk add build-base` and re-using the `GOPATH` cache. Also the nesting goes one container deeper.

> It seems like we only want to `docker run -v lifecycle-out:/lifecycle --rm $(SOURCE_COMPILATION_IMAGE) tar -cf- out | tar -xf-` after particular tasks?

Good call, I feel like the way we implement `build-linux` (above) might dictate how we want to pull artifacts out of these containers (or nested containers if we go with DIND). I can do a follow-up pass to take care of both. Would we be

Reply:
> It seems like `docker-run-linux` has some issues. It can't run `make acceptance` b/c the image is missing the `docker` CLI.

Right, I believe that's due to the `make docker-run-linux` (container)-> `make acceptance` -> `go test acceptance_test.go` -> `make build-linux` -> `docker run ... go build` chain. I'm interested in how this could best work:

1) Changing `make build-linux` to be more like - though not quite equal to - `make build-windows` (which now switches based on OS to either `build-windows-on-windows` or `build-windows-on-posix`). 

    Maybe the targets for `build-linux` could be something like `build-linux-on-linux`/`build-linux-on-docker` depending on OS. The target `build-linux-on-docker` would essentially just do `docker-run-linux DOCKER_CMD="make build-linux-on-linux"`. Repeated `build-linux-on-docker` runs would be faster since we could remove the `apk add build-base` for every call and GOPATH could be cached.
2) or, adding `docker` to the `tools/Dockerfile.linux` image to support DIND, though we'd miss out on speedup from of removing `apk add build-base` and re-using the `GOPATH` cache. Also the nesting goes one container deeper.

> It seems like we only want to `docker run -v lifecycle-out:/lifecycle --rm $(SOURCE_COMPILATION_IMAGE) tar -cf- out | tar -xf-` after particular tasks?

Good call, I feel like the way we implement `build-linux` (above) might dictate how we want to pull artifacts out of these containers (or nested containers if we go with DIND). I can do a follow-up pass to take care of both, depending on what is preferred.

Reply:
Actually, how about we delete the `docker-run-linux` target for now and I'll open a separate PR for it.